c++/src/kj/buffering.c++

```diff
@@ -0,0 +1,350 @@
+#include "buffering.h"
+#include <kj/debug.h>
+
+namespace kj {
+
+namespace _ {
+  static inline uint lg(uint value) {
+    // Compute floor(log2(value)).
+    //
+    // Undefined for value = 0.
+  #if _MSC_VER && !defined(__clang__)
+    unsigned long i;
+    auto found = _BitScanReverse(&i, value);
+    KJ_DASSERT(found);  // !found means value = 0
+    return i;
+  #else
+    return sizeof(uint) * 8 - 1 - __builtin_clz(value);
+  #endif
+  }
+}
+
+RetainableFdBufferRange::RetainableFdBufferRange(Badge<SplittableFdBufferRange>,
+    FdRangeInputBuffer& parent, ArrayPtr<byte> bytes, Array<AutoCloseFd> fds)
+    : parentOrRetainedBytes(&parent),
+      bytes(bytes),
+      fds(mv(fds)) {}
+
+RetainableFdBufferRange::RetainableFdBufferRange(Badge<SplittableFdBufferRange>,
+    Own<void> ownBytes, ArrayPtr<byte> bytes, Array<AutoCloseFd> fds)
+    : parentOrRetainedBytes(mv(ownBytes)),
+      bytes(bytes),
+      fds(mv(fds)) {}
+
+ArrayPtr<byte> RetainableFdBufferRange::getBytes() { return bytes; }
+ArrayPtr<AutoCloseFd> RetainableFdBufferRange::getFds() { return fds; }
+
+void RetainableFdBufferRange::retain() {
+  KJ_SWITCH_ONEOF(parentOrRetainedBytes) {
```

Comment:
Okay, upon review, I still feel like there's a world where we can avoid large copies altogether. I notice we're using `FlatArrayMessageReader` in serialize-async.c++ but we don't have to be. If you read a message into two refcounted buffers, you can stitch them together into one message reader. (You might need to copy one segment of the message into a smaller "transition" buffer.)

I'm pushing on this because both allocating a large buffer and copying a large amount of data can be a "stop the world" moment for low level IO code. Anything we do to reduce that will be worth it.